clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -3100,8 +3100,13 @@ raw2trace_t::write(raw2trace_thread_data_t *tdata, const trace_entry_t *start,
             }
```

Comment:
For L3082: It'd be useful to add a comment above that reminds the reader that prev_was_encoding is to avoid a duplicate encoding being emitted after an actual encoding entry, specifically for the case *after we open a new chunk* (because record_encoding_emitted will still return `true` since we just cleared that cache after opening a new chunk)

(or possibly a better worded summary)

Reply:
Hmm: your questions are making me realize that record_encoding_emitted is never called for these delayed encodings (it was called pre-chunk but was cleared) -- it seems like what we want is to call that, and get rid of prev_was_encoding.  I just added a 2nd ret to the unit test and indeed we (incorrectly) emit another encoding there!

Reply:
> your questions are making me realize that record_encoding_emitted is never called for these delayed encodings (it was called pre-chunk but was cleared) 

Wait why?

By delayed encoding, you mean an encoding that we deliberately repeat after opening a new chunk (L3091), right? But control would enter that if-block only after calling record_encoding_emitted.

Reply:
That if-block is only for instructions and it short-circuits if prev_was_encoding.  But we don't have the hash key: so I think we have to wait for the instr and if prev_was_encoding add to the hash there.

Reply:
> That if-block is only for instructions

Right, but that's fine because we want to emit the delayed encoding only when we see an actual instruction for it.

> it short-circuits if prev_was_encoding

Right, but for delayed encodings, by definition, we don't have a new encoding just before that instruction (we had an encoding before a previous dynamic instance of that instruction).

Reply:
> Right, but for delayed encodings, by definition, we don't have a new encoding just before that instruction (we had an encoding before a previous dynamic instance of that instruction).

This *is* the previous dynamic instance: the delayed branch.

I've put in the updated test and new code if you'd like to take a look.

Reply:
Going to go ahead with this.